src/Google.Api.Gax/TemplatedResourceName.cs

```diff
@@ -107,31 +106,49 @@ public ResourceName(PathTemplate template, params string[] resourceIds)
         /// Creates a clone of this resource name, which is then independent of the original.
         /// </summary>
         /// <returns>A clone of this resource name.</returns>
-        public ResourceName Clone()
+        public TemplatedResourceName Clone()
         {
-            return new ResourceName(Template, ServiceName, (string[]) _resourceIds.Clone(), true);
+            return new TemplatedResourceName(Template, ServiceName, (string[]) _resourceIds.Clone(), true);
         }
 
         /// <summary>
         /// Private constructor used by internal code to avoid repeated cloning and validation.
         /// </summary>
-        private ResourceName(PathTemplate template, string serviceName, string[] resourceIds, bool ignored)
+        private TemplatedResourceName(PathTemplate template, string serviceName, string[] resourceIds, bool ignored)
         {
             Template = GaxPreconditions.CheckNotNull(template, nameof(template));
             ServiceName = serviceName;
             this._resourceIds = resourceIds;
         }
 
-        internal static ResourceName CreateWithShallowCopy(PathTemplate template, string serviceName, string[] resourceIds)
+        internal static TemplatedResourceName CreateWithShallowCopy(PathTemplate template, string serviceName, string[] resourceIds)
         {
-            return new ResourceName(template, serviceName, resourceIds, true);
+            return new TemplatedResourceName(template, serviceName, resourceIds, true);
         }
 
+        /// <inheritdoc />
+        public ResourceNameKind Kind => ResourceNameKind.Templated;
+
         /// <summary>
         /// Returns a string representation of this resource name, expanding the template
         /// parameters with the resource IDs and prepending the service name (if present).
         /// </summary>
         /// <returns>A string representation of this resource name.</returns>
         public override string ToString() => Template.ReplaceParameters(ServiceName, _resourceIds);
+
+        /// <inheritdoc />
+        public override int GetHashCode() => ToString().GetHashCode();
+
+        /// <inheritdoc />
+        public override bool Equals(object obj) => Equals(obj as TemplatedResourceName);
+
+        /// <inheritdoc />
+        public bool Equals(TemplatedResourceName other) => ToString() == other?.ToString();
```

Comment:
Hmm. I wonder whether we should check the template as well. Let's not worry for now.